proposals/0064-transaction-receipt.md

```diff
@@ -0,0 +1,257 @@
+  [TransactionResult]: https://docs.rs/solana-sdk/1.16.1/solana_sdk/transaction/type.Result.html
+
+### Proof-of-History / Block Hash
+
+Another alternative to introducing a new commitment scheme is reusing the
+proof-of-history (PoH) hash chain. This was proposed in
+[SPV Proposal](https://docs.solana.com/proposals/simple-payment-and-state-verification#transaction-merkle).
+
+The PoH chain currently commits to the signatures of all transactions added to
+the ledger. The consensus layer then periodically votes on the last state of the
+PoH chain for each bankhash which is an extension of the blockhash.
+Expanding the PoH hash is the least complex option as of today but
+is consequential for future upgrades.
+
+Such a change would be incompatible with design goal 2 because it redefines
+the PoH hash to additionally commit to execution results, instead of only
+ledger content. Block producers are then forced to synchronously replay while
+appending PoH ticks.
+
+Furthermore, it significantly changes behavior in the event of an execution
+disagreement (e.g. due to a difference in execution behavior between
+validators). Mixing execution results into PoH forces execution disagreements to
+result in a chain split.
+
+## Detailed Design
+
+### Transaction Receipt Specification
+
+The transaction receipt must contain the following information related to the transaction:
+
+- Message Hash
+- Execution Status
+
+The receipt would be a structure defined as:
+
+```rust
+pub struct TransactionReceipt {
+    pub message_hash: [u8;32],
+    pub status: u8 // 1 or 0 would determine the post execution status
+}
+```
+
+### Receipt Tree Specification
+
+The receipt tree is a binary merkle tree of receipts, where each node is a 32
+byte hash of the Receipt data structure.
+
+We construct a deterministic tree over a list of Receipts per slot with
+the following properties:
+
+- The tree needs to be strictly deterministic as any other cryptographic
+  primitive to ensure that trees are exactly identical when individually
+  constructed by different nodes.
+
+- The order of the leaves should match the order of the list
+  of receipts. When validators replay the transactions in a slot, they should aggregate
+  the receipts in the same order as the transactions were in the block.
+
+- For membership proofs and inclusion checks one should be
+  able to provide a path from the leaf node (Receipt) to the root of the tree.
+  The locally computed root is compared for equality.
+  
+- Finally after aggregating all the receipts and constructing the final root hash,
+  the count of the receipts is hashed with the root in 64-bit little endian byte
+  ordering to produce a final commitment hash. This is done to prevent the
+  possibility of length extension attack vector inherent to merkle trees
+  where the total number of leaves is not fixed/known before tree construction.
```

Comment:
> Can you please clarify whether the root of the empty set is `sha256(concat(sha256([])), 0u64)` or `sha256(0u64)`?

it would be the latter but with the prefix we just discussed about: `sha256(concat(0x80,0u64))`

Reply:
got it